DirectProgramming/C++SYCL_FPGA/Tutorials/Features/experimental/ip_authoring_mmhost/src/mmhost.cpp

```diff
@@ -0,0 +1,185 @@
+
+using namespace sycl;
+using ext::intel::experimental::property::usm::buffer_location;
+
+template <typename T>
+inline std::shared_ptr<T> make_malloc_shared(queue &q, int n, int kBL) {
+  T *mem =
+      malloc_shared<T>(sizeof(T) * n, q, property_list{buffer_location(kBL)});
+  return std::move(
+      std::shared_ptr<T>(mem, [&q](T *ptr) { sycl::free(ptr, q); }));
+}
+
+constexpr int BL1 = 1;
+constexpr int BL2 = 2;
+constexpr int BL3 = 3;
+
+// Using pointers,
+struct PointerIP {
+  int *x, *y, *z;
+  int size;
+
+  PointerIP(int *x_, int *y_, int *z_, int size_)
+      : x(x_), y(y_), z(z_), size(size_) {}
+
+  void operator()() const {
+    for (int i = 0; i < size; ++i) {
+      int mul = x[i] * y[i];
+      int add = mul + z[i];
+      y[i] = mul;
+      z[i] = add;
+    }
+  }
+};
+
+// Using mmhost macro
+struct VectorMADIP {
+  mmhost(BL1,  // buffer_location or aspace
+         28,   // address width
+         64,   // data width
+         16,   // latency
+         1,    // read_write_mode, 0: ReadWrite, 1: Read, 2: Write
+         1,    // maxburst
+         0,    // align, 0 defaults to alignment of the type
+         1     // waitrequest, 0: false, 1: true
+         ) int *x;
+  mmhost(BL2,  // buffer_location or aspace
+         28,   // address width
+         64,   // data width
+         16,   // ! latency, must be atleast 16
+         0,    // read_write_mode, 0: ReadWrite, 1: Read, 2: Write
+         1,    // maxburst
+         0,    // align, 0 defaults to alignment of the type
+         1     // waitrequest, 0: false, 1: true
+         ) int *y;
+  mmhost(BL3,  // buffer_location or aspace
+         28,   // address width
+         64,   // data width
+         16,   // ! latency, must be atleast 16
+         0,    // read_write_mode, 0: ReadWrite, 1: Read, 2: Write
+         1,    // maxburst
+         0,    // align, 0 defaults to alignment of the type
+         1     // waitrequest, 0: false, 1: true
+         ) int *z;
+  int size;
+
+  VectorMADIP(int *x_, int *y_, int *z_, int size_)
+      : x(x_), y(y_), z(z_), size(size_) {}
+
+  void operator()() const {
+    for (int i = 0; i < size; ++i) {
+      int mul = x[i] * y[i];
+      int add = mul + z[i];
+      y[i] = mul;
+      z[i] = add;
+    }
+  }
+};
+
+void fillArrays(int *x, int *y, int *z, int size) {
+  for (int i = 0; i < size; ++i) {
+    x[i] = i;
+    y[i] = i * 2;
+    z[i] = i * 3;
+  }
+}
+
+int main(void) {
+#if FPGA_SIMULATOR
+  auto selector = sycl::ext::intel::fpga_simulator_selector_v;
+#elif FPGA_HARDWARE
+  auto selector = sycl::ext::intel::fpga_selector_v;
+#else  // #if FPGA_EMULATOR
+  auto selector = sycl::ext::intel::fpga_emulator_selector_v;
+#endif
+
+  try {
+    sycl::queue q(selector, fpga_tools::exception_handler,
+                  property::queue::enable_profiling{});
+
+    // Print out the device information.
+    std::cout << "Running on device: "
+              << q.get_device().get_info<info::device::name>() << "\n";
+
```

Comment:
This design depends on USM Host support (due to the `malloc_shared` calls). It might be a best practice to check that USM host/shared support exists. 

```suggestion
    // make sure the device supports USM host allocations
    if (!device.has(sycl::aspect::usm_host_allocations)) {
      std::cerr << "This design must either target a board that supports USM "
                   "Host/Shared allocations, or IP Component Authoring. "
                << std::endl;
      std::terminate();
    }
```